### ISSUE TITLE
Color syntax coloring changes after first execution

### ISSUE THREAD

**Issue by dsblank** [ipython/ipython#7635]
In master, using a wrapper kernel, language is "java". Here is a copy/paste of kernel_info in JS console:

```
IPython.notebook.kernel.info_reply
 {protocol_version: "5.0", help_links: Array[1], 
  implementation_version: "1.0", 
  language_info: { 
             help_links: Array[1]
             implementation: "Processing"
             implementation_version: "1.0"
             extension: ".pjs"
             mimetype: "text/x-processing"
             name: "java"
             protocol_version: "5.0"
}
```
`IPython.notebook.codemirror_mode` is still "java", but as you can see it doesn't highlight in second cell:

![codemirror](https://cloud.githubusercontent.com/assets/168568/5978855/02bc8d9a-a871-11e4-8901-1b1b0ff07575.png)

It loads fine, but after the first cell evaluation, it is broken. Why other information would help debug this?

**Comment by minrk** [comment]
Hm, I thought we had sorted this out. Does it affect only new cells? Or are new cells correct?

**Comment by dsblank** [comment]
I know what causes it: a badly formed CodeMirror mode name: it was `java` rather than `text/x-java`... I'll check to see what is causing this...

**Comment by dsblank** [comment]
So I had a language_info defined like:

```python
    language_info = {
        'mimetype': 'text/x-java',
        'name': 'java',
        'file_extension': '.java',
    }
```
but needed to change it to:

```python
    language_info = {
        'mimetype': 'text/x-java',
        'name': 'java',
        'codemirror_mode': {
            "version": 2,
            "name": "text/x-java"
        },
        'file_extension': '.java',
    }
 ```

in order for formatting to work. I don't know if this is a bug or not.

**Comment by minrk** [comment]
I think only language_info.name is used to look up codemirror mode in the notebook, and it seems like we may want to use the mimetype as well. The text editor looks up by mimetype, so perhaps we should copy some of that over.

**Comment by minrk** [comment]
I think we *should* be looking at the mime-type in language_info in the notebook like we already do in the text editor, but that requires a bit of reorganization in the notebook js that I don't want to do this close to 3.0. Marking as 3.1, and kernels where language.name doesn't correspond to a codemirror mode can use the codemirror_mode field in the meantime.

**Comment by Carreau** [comment]
I think this one has been fixed in 3.2 o 3.x
